packages/starlight/utils/virtual-page.ts

```diff
@@ -0,0 +1,169 @@
+	 * collection entries.
+	 */
+	id: string;
+};
+
+export function generateVirtualRouteData({
+	props,
+	url,
+}: {
+	props: VirtualPageProps;
+	url: URL;
+}): StarlightRouteData {
+	const { isFallback, frontmatter, slug, ...routeProps } = props;
+	const virtualFrontmatter = getVirtualFrontmatter(frontmatter);
+	const id = `${stripLeadingAndTrailingSlashes(slug)}.md`;
+	const localeData = slugToLocaleData(slug);
+	const sidebar = props.sidebar ?? getSidebar(url.pathname, localeData.locale);
+	const headings = props.headings ?? [];
+	const virtualEntry: VirtualDocsEntry = {
+		id,
+		slug,
+		body: '',
+		collection: 'docs',
+		data: {
+			...virtualFrontmatter,
+			sidebar: {
+				attrs: {},
+				hidden: false,
+			},
+		},
+	};
+	const entry = virtualEntry as StarlightDocsEntry;
+	const entryMeta: StarlightRouteData['entryMeta'] = {
+		dir: props.dir ?? localeData.dir,
+		lang: props.lang ?? localeData.lang,
+		locale: localeData.locale,
+	};
+	const editUrl = virtualFrontmatter.editUrl ? new URL(virtualFrontmatter.editUrl) : undefined;
+	const lastUpdated =
+		virtualFrontmatter.lastUpdated instanceof Date ? virtualFrontmatter.lastUpdated : undefined;
+	const routeData: StarlightRouteData = {
+		...routeProps,
+		...localeData,
+		id,
+		editUrl,
+		entry,
+		entryMeta,
+		hasSidebar: props.hasSidebar ?? entry.data.template !== 'splash',
+		headings,
+		labels: useTranslations(localeData.locale).all(),
+		lastUpdated,
+		pagination: getPrevNextLinks(sidebar, config.pagination, entry.data),
+		sidebar,
+		slug,
+		toc: getToC({
+			...routeProps,
+			...localeData,
+			entry,
+			entryMeta,
+			headings,
+			id,
+			locale: localeData.locale,
+			slug,
+		}),
+	};
+	if (isFallback) {
+		routeData.isFallback = true;
+	}
+	return routeData;
+}
+
+/** Validates the virtual frontmatter properties from the props received by a virtual page. */
+function getVirtualFrontmatter(frontmatter: StarlightVirtualFrontmatter) {
+	// This needs to be in sync with ImageMetadata.
+	// https://github.com/withastro/astro/blob/cf993bc263b58502096f00d383266cd179f331af/packages/astro/src/assets/types.ts#L32
+	return StarlightVirtualFrontmatterSchema({
+		image: () =>
+			z.object({
+				src: z.string(),
+				width: z.number(),
+				height: z.number(),
+				format: z.union([
+					z.literal('png'),
+					z.literal('jpg'),
+					z.literal('jpeg'),
+					z.literal('tiff'),
+					z.literal('webp'),
+					z.literal('gif'),
+					z.literal('svg'),
+					z.literal('avif'),
+				]),
+			}),
+	}).parse(frontmatter);
+}
+
+// https://stackoverflow.com/a/66252656/1945960
+type RemoveIndexSignature<TType> = {
+	[Tkey in keyof TType as string extends Tkey
+		? never
+		: number extends Tkey
+		? never
+		: symbol extends Tkey
+		? never
+		: Tkey]: TType[Tkey];
+};
+
+// https://www.totaltypescript.com/concepts/the-prettify-helper
+type Prettify<TType> = {
+	[TKey in keyof TType]: TType[TKey];
+} & {};
```

Comment:
This is totally a stylistic nit, but we don’t generally prefix type names with `T` in other code. I think this could even be as simple as:

```suggestion
// https://stackoverflow.com/a/66252656/1945960
type RemoveIndexSignature<T> = {
	[K in keyof T as string extends K
		? never
		: number extends K
		? never
		: symbol extends K
		? never
		: K]: T[K];
};

// https://www.totaltypescript.com/concepts/the-prettify-helper
type Prettify<T> = {
	[K in keyof T]: T[K];
} & {};
```

Reply:
Or if we prefer being a bit less terse, we could use the full words:

```suggestion
// https://stackoverflow.com/a/66252656/1945960
type RemoveIndexSignature<Type> = {
	[Key in keyof Type as string extends Key
		? never
		: number extends Key
		? never
		: symbol extends Key
		? never
		: Key]: Type[Key];
};

// https://www.totaltypescript.com/concepts/the-prettify-helper
type Prettify<Type> = {
	[Key in keyof Type]: Type[Key];
} & {};
```